#20 rectangle of interest

dustin-graham opened the pull request:
# Fixes #20 Specify a rectangle of interest which limits barcode scanning

_**NOTE: THIS IS BASED ON PR #55 AND #57 , THOSE SHOULD BE MERGED BEFORE THIS ONE**_

<img src="https://user-images.githubusercontent.com/959931/149459914-823f7c14-03fb-4398-9302-60207b4f7416.jpeg" width="250"/>

This PR updates the MaterialFinderOverlay so that it can use it's finder rectangle to filter codes scanned by the platform.

## Filtering codes within the "rectangle of interest"
Prior to this PR, the MaterialFinderOverlay was mostly cosmetic. It responded to any codes being scanned within the whole image and not specifically to those being scanned within the finder rectangle. The MaterialFinderOverlay now requires a "rectangle of interest" so that it can filter through all the codes scanned and only return those that are completely contained within the rectangle.

## Optional Scan Line
I also added an optional "scan line" that can be drawn horizontally through the middle of the box. I found that on iOS, AVFoundation has trouble in some cases recognizing linear codes like Code128 unless the code is in the middle of the screen. This can be confusing to the end user without any sort of indicator. The scan line is a visual metaphor of familiar laser barcode scanners. This should prompt the user to position the linear barcode on the line. However, not all applications will find this useful so it is left as an option and disabled by default.


## Barcode aware finder border
I also added a callback that allows users to customize the color of the finder border paint based on the barcode(s) scanned. In my use case, I need to show "green" when the user scans a code we are expecting and "red" when the code is not valid. This gives us the flexibility we need to give feedback to our users.

dustin-graham committed:
Android code coordinate implementation

dustin-graham committed:
ios code coordinate implementation

dustin-graham committed:
cleanup

dustin-graham committed:
resolve analysis warnings

dustin-graham committed:
fix method channel test

dustin-graham committed:
#56 return all scanned codes

dustin-graham committed:
fix formatting of PreviewConfiguration.kt

dustin-graham committed:
fix formatting of PreviewConfiguration.kt

dustin-graham committed:
fix method channel test (again)

accidentally included code from another branch that broke the test

dustin-graham committed:
remove analysis string in message channel

clean up kotlin code style

dustin-graham committed:
make overlay configurable

dustin-graham committed:
improve class naming

dustin-graham committed:
move coordinate conversion to swift plugin

dustin-graham committed:
make PreviewView not static

dustin-graham committed:
improve file and class naming

dustin-graham committed:
remove rxdart dependency

dustin-graham committed:
Merge branch '52_code_coordinates' into 56_all_scanned_codes

# Conflicts:
#	fast_barcode_scanner/android/src/main/kotlin/com/jhoogstraat/fast_barcode_scanner/types/PreviewConfiguration.kt
#	fast_barcode_scanner/ios/Classes/Scanner/AVFoundationBarcodeScanner.swift
#	fast_barcode_scanner/lib/src/camera_controller.dart
#	fast_barcode_scanner_platform_interface/test/method_channel_fast_barcode_scanner_test.dart

dustin-graham committed:
improve variable naming in swift plugin

dustin-graham committed:
#20 add rectangle of interest

the rectangle of interest functionality is built into the
MaterialFinderOverlay. Provide the overlay with a rectangle
of interest and a filtered code callback.

dustin-graham committed:
Merge branch 'develop' into 56_all_scanned_codes

# Conflicts:
#	fast_barcode_scanner/android/src/main/kotlin/com/jhoogstraat/fast_barcode_scanner/FastBarcodeScannerPlugin.kt
#	fast_barcode_scanner/example/lib/scanning_screen/scanning_screen.dart
#	fast_barcode_scanner/ios/Classes/FastBarcodeScannerPlugin.swift
#	fast_barcode_scanner/ios/Classes/Scanner/AVFoundationBarcodeScanner.swift
#	fast_barcode_scanner/lib/src/barcode_camera.dart
#	fast_barcode_scanner/lib/src/camera_controller.dart
#	fast_barcode_scanner/lib/src/overlays/code_boundary_overlay/code_border_painter.dart
#	fast_barcode_scanner/lib/src/overlays/code_boundary_overlay/code_boundary_overlay.dart
#	fast_barcode_scanner/pubspec.yaml

dustin-graham committed:
comment and variable name cleanup

dustin-graham committed:
Merge branch '56_all_scanned_codes' into 20_rectangle_of_interest

jhoogstraat commented:
Now that #55 and #57 are merged, we can tackle this pr next.
Really awesome how you brought the project forward @dustin-graham ! Much appreciated.

dustin-graham committed:
Merge branch 'develop' into 20_rectangle_of_interest

# Conflicts:
#	fast_barcode_scanner/lib/src/camera_controller.dart
#	fast_barcode_scanner_platform_interface/lib/src/fast_barcode_scanner_platform_interface.dart

dustin-graham commented:
@jhoogstraat , merged with `develop`. Ready for review.

dustin-graham committed:
run dart formatter

dustin-graham committed:
make paint static

dustin-graham commented:
> Now that #55 and #57 are merged, we can tackle this pr next. Really awesome how you brought the project forward @dustin-graham ! Much appreciated.

You are very welcome! This project already has a very solid foundation. It's my pleasure to help move it forward.

jhoogstraat commented:
The MaterialBarcodeOverlay is aming to replicate the barcode scanner outlined [here](https://material.io/design/machine-learning/barcode-scanning.html#components). 
Maybe we can separate the red line into its own overlay?

Edit: just saw that you can toggle it. Not a big fan of it visually inside the cut out though.

jhoogstraat commented on fast_barcode_scanner/lib/src/camera_controller.dart:
```diff
@@ -154,27 +154,33 @@ class _CameraController implements CameraController {
   /// User-defined handler, called when a barcode is detected
   OnDetectionHandler? _onScan;
 
+  OnDetectionHandler _buildScanHandler(OnDetectionHandler? onScan) {
```
Small doc mentioning that this is a curried function for onScan.

dustin-graham replied:
will fix.

jhoogstraat commented on fast_barcode_scanner/lib/src/overlays/material_finder_overlay/material_finder_overlay.dart:
```diff
@@ -1,10 +1,12 @@
-import '../../camera_controller.dart';
-import '../../types/scanner_event.dart';
+import 'package:fast_barcode_scanner/fast_barcode_scanner.dart';
```
Please keep local path imports.

dustin-graham replied:
will fix.

jhoogstraat commented on fast_barcode_scanner/lib/src/overlays/rect_of_interest/rect_of_interest.dart:
```diff
@@ -0,0 +1,129 @@
+  }
+}
+
+class WideRectOfInterest extends RectOfInterest {
```
This is not great for discoverability.
It would be nice to be able to do something like `RectOfInterest.wide()`.

Maybe also a rect of interest that creates itself dynamically according to the code types that are gonna be scanned?

dustin-graham replied:
I like the discoverability suggestion. I'm adding that.

However, I think I'd rather leave the selection of square or rectangle to the user.

jhoogstraat commented on fast_barcode_scanner/lib/src/overlays/rect_of_interest/rect_of_interest.dart:
```diff
@@ -0,0 +1,129 @@
+
+  Rect rect(Rect previewWidgetRect);
+
+  CodeFilter codeFilter({
```
I cannot check currently, if this method is exposed to the user, but it should be called something like `_buildCodeFilter` as it's a curried function aswell.

dustin-graham replied:
It is used by `MaterialFinderOverlay`. Great naming suggestion. Will fix.

jhoogstraat reviewed (commented):
Really pushing the boundaries of the original idea of "overlays" which were meant to be cosmetic.

I am not 100% sure if having multiple `onScan`s is the was to go. This may lead to potential errors if the user is not careful.

dustin-graham commented:
> Really pushing the boundaries of the original idea of "overlays" which were meant to be cosmetic.
> 
> I am not 100% sure if having multiple `onScan`s is the was to go. This may lead to potential errors if the user is not careful.

I understand the concern. I went through multiple iterations on this design and this is the one I settled on as the best I could think of. 

On the topic of cosmetics:
The MaterialBarcodeOverlay is actually meant to be more than just cosmetic. If you look at the animation in the "Barcode Frame" section of the [doc](https://material.io/design/machine-learning/barcode-scanning.html#components), you will see that the barcode does not scan until it is within the bounds of the frame. The frame is meant to filter the visible codes.

Multiple `onScan`'s:
My first implementation had the main `onScan` only returning codes that were within the frame. However, the implementation ended up being very messy. Currently, the overlay system is appropriately decoupled from the BarcodeCamera. The overlays that the user chooses to include (including their own if they like) are arbitrary. The BarcodeCamera and the CameraController don't know the specific details of any of the overlays and every component has its own dedicated job. This is good. However, to get a single onScan function to effectively filter the codes according to the constraints of an arbitrary overlay, I had to introduce various systems that ultimately added coupling between these components. I implemented this 3 or 4 different ways. This requires coordination between the visual size of the frame and the coordinates of the scanned codes. The best of these ways included adding a 1st-class concept for a "filtering overlay" interface which the base components could be aware of and respect when receiving codes from the platform interface. But this seemed overly-contrived/AdHoc. So while having a single `onScan` could be considered a step forward in our public API, I think it's several steps backwards in architecture.

Besides the architecture difficulties, I thought about this for a long while in terms of the single-responsibility principle. The BarcodeCamera's job is to scan all the barcodes that it can clearly see. The purpose of the material overlay is to filter out visible codes and only scan those that are completely within its bounds. This is its job, it's single responsibility. If the user wants to filter codes by location on the screen, they add a component designed to do just that, and a filtered callback for just those codes is within _that_ component's area of responsibility, not the BarcodeCamera.

This was roughly the story of how I worked through the problem and ultimately ended up with this current design.

dustin-graham commented:
> The MaterialBarcodeOverlay is aming to replicate the barcode scanner outlined [here](https://material.io/design/machine-learning/barcode-scanning.html#components).
> Maybe we can separate the red line into its own overlay?
> 
> Edit: just saw that you can toggle it. Not a big fan of it visually inside the cut out though.

So the main use case for including the red "scan line" is when the user is on iOS and we are using AVFoundation for the scanning. AVFoundation has documented limitations for scanning linear codes. It will only recognize them if they are exactly in the middle of the screen. This limitation is not obvious to end users. The "Scan line" may provide enough of a hint for the end user to move the code into the middle of the screen so it can be scanned. Vision does not have this limitation. 

I recognize that some may not find this line visually appealing. For this reason I initially intended to add this as a separate overlay but I found it difficult to coordinate multiple overlays so that the line was exactly inside the cutout. It was most convenient to put it in the same overlay and make it optional. Plus, as of _this_ PR, we only support AVFoundation out of the box and so this scan line can be super useful for end users. Once we get the ability to choose Vision, this becomes a weaker argument.

If you were to insist that we remove it, I wouldn't put up a fight. I would just need to find another way to add it in my app without duplicating code. I'm confident I can do it. Please consider and let me know what you think.

dustin-graham committed:
respond to PR feedback.

dustin-graham commented:
@jhoogstraat , thank you for your thorough code review. I responded to each item and pushed updated code with requested fixes.

I also left some comments here in this PR to provide justifications for several design decisions you noticed. Hopefully these help clarify things. To summarize:

1. Material overlay having it's own `onScan` callback: I think this is the best way to support the single-responsibility principle and keep BarcodeCamera, CameraController, and any included overlays as decoupled as possible.
2. Red Scan Line: This is a very useful guide to help iOS users scanning with AVFoundation while scanning barcodes because they only scan when exactly in the middle of the screen. It's included in the Material Finder because their purposes seem to overlap and it made the implementation easier. I'm willing to move it out if you insist, but I think it's a valuable addition while we support AVFoundation.

jhoogstraat commented:
Thanks for your detailed explanation! 
You wrote
> The BarcodeCamera's job is to scan all the barcodes that it can clearly see. The purpose of the material overlay is to filter out visible codes and only scan those that are completely within its bounds. This is its job, it's single responsibility. If the user wants to filter codes by location on the screen, they add a component designed to do just that, and a filtered callback for just those codes is within that component's area of responsibility, not the BarcodeCamera.

and i completely agree with that. I think that really elevates the possibilities and customizability of the scanner. 
Its even necessary if we want to implement the material barcode scanner fully (as you mentioned).

Regarding the scan line:
I still think its not very appealing inside the material overlay. Moreover, the cutout is already prompting the user to align the barcode inside of it.
However, the scan line is certainly connected with traditional barcode scanning and I feel we should provide it in a separate overlay. That way the user can choose between different ui styles. One problem that arises is that the barcode filtering would be duplicated in both overlays...

jhoogstraat commented:
Ping @dustin-graham

dustin-graham commented:
@jhoogstraat , thanks for the patience. It's been busy at work the past couple weeks. I'm taking care of the remaining feedback today.

dustin-graham committed:
remove scan line UI

This is a non-essential feature that nobody has requested.
It's easy enough for anyone to add their own custom overlay to do this.

dustin-graham commented:
@jhoogstraat , ready for review again. I decided to just remove the scan line UI. It was easy enough to make a cosmetic overlay that just shows the line but does not filter codes. However, it proved to be much more difficult to filter codes  with a single line. Linear codes like code128 don't get consistent corner points around the barcode making it difficult to reliably measure when the code intersects the line. And since the primary utility of the line was to guide users when scanning linear codes while using AVFoundation it doesn't have a lot of utility if it can't reliably intersect with linear codes. It has some utility as a cosmetic visual guide, but I think it might be best to leave something like this for the user to implement themselves as a custom overlay. Additionally, nobody else has asked for this functionality. And so for the sake of simplicity and expediency, I decided to just remove it from this PR.

dustin-graham commented:
@jhoogstraat ping^

CoryADavis commented:
@dustin-graham Really exciting features you have developed! I've been eagerly following these PRs.
@jhoogstraat Thanks for the excellent library.

dustin-graham commented:
@jhoogstraat , with the removal of the scanning line. Are you comfortable if I merge this PR?

jhoogstraat commented:
Sorry for being absent, but I have still a lot of things going on (more then expected). I am sure the pr is in very good state to get merged. Please go ahead!

tarang1998 commented:
@jhoogstraat @dustin-graham 

When can we expect this feature to be published in pub

I tried to build my app with the source code of the "develop" branch but currently am facing some issues 

Thanks a lot for your work :)
